folding-schemes/src/folding/hypernova/mod.rs

```
@@ -25,3 +55,535 @@ impl<F: PrimeField> Witness<F> {
+
+        let r1cs_z = [
+            vec![C1::ScalarField::one()],
+            r1cs_x_i1.clone(),
+            r1cs_w_i1.clone(),
+        ]
+        .concat();
+        // compute committed instances, w_{i+1}, u_{i+1}, which will be used as w_i, u_i, so we
+        // assign them directly to w_i, u_i.
+        let (u_i, w_i) = self
+            .ccs
+            .to_cccs::<_, C1, CS1>(&mut rng, &self.cs_params, &r1cs_z)?;
+        self.u_i = u_i.clone();
+        self.w_i = w_i.clone();
+
+        // set values for next iteration
+        self.i += C1::ScalarField::one();
+        // assign z_{i+1} into z_i
+        self.z_i = z_i1.clone();
+        self.U_i = U_i1.clone();
+        self.W_i = W_i1.clone();
+
+        #[cfg(test)]
+        {
+            // check the new LCCCS instance relation
+            self.U_i.check_relation(&self.ccs, &self.W_i)?;
+            // check the new CCCS instance relation
+            self.u_i.check_relation(&self.ccs, &self.w_i)?;
+        }
+
+        Ok(())
+    }
+
+    fn state(&self) -> Vec<C1::ScalarField> {
+        self.z_i.clone()
+    }
+
+    fn instances(
+        &self,
+    ) -> (
+        Self::RunningInstance,
+        Self::IncomingInstance,
+        Self::CFInstance,
+    ) {
+        (
+            (self.U_i.clone(), self.W_i.clone()),
+            (self.u_i.clone(), self.w_i.clone()),
+            (self.cf_U_i.clone(), self.cf_W_i.clone()),
+        )
+    }
+
+    /// Implements IVC.V of HyperNova+CycleFold. Notice that this method does not include the
+    /// commitments verification, which is done in the Decider.
+    fn verify(
+        vp: Self::VerifierParam,
+        z_0: Vec<C1::ScalarField>, // initial state
+        z_i: Vec<C1::ScalarField>, // last state
+        num_steps: C1::ScalarField,
+        running_instance: Self::RunningInstance,
+        incoming_instance: Self::IncomingInstance,
+        cyclefold_instance: Self::CFInstance,
+    ) -> Result<(), Error> {
```

We might want to also include the case where $i == 0$, which consists in checking that $z_0 == z_i$

---

that's a good point, in fact, I'd rather return an error as in the Decider when i==0 in the verification (I think that it was you who proposed to return the error in the Decider::verify when i<=1). Updated adding the if<=1 return err

---

I don't remember, I may have said this, but I don't think this is correct. I think it would be better to just follow what the paper says regarding the verifier $\mathcal{V}$, no?